From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 24 Apr 2025 16:36:16 -0400
Subject: [PATCH] Paw optimization

Some random optimizations

- Remove Paper's dead code
- Only set shuffle random seed if is really used
- Cache direction values to skip copy
- Secret patches (WIP)

diff --git a/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java b/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
index 3b889d80b1b0f2b3351595c1005a4b3cb538d4dd..25ed154b76b7745f668d0ae2bf57802f22f183b8 100644
--- a/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaAdventureAdvancements.java
@@ -1104,7 +1104,7 @@ public class VanillaAdventureAdvancements implements AdvancementSubProvider {
     private static Criterion<ItemUsedOnLocationTrigger.TriggerInstance> placedBlockActivatesCreakingHeart(
         HolderGetter<Block> blockRegistry, TagKey<Block> blockTag
     ) {
-        LootItemCondition.Builder[] builders = Stream.of(Direction.values())
+        LootItemCondition.Builder[] builders = Stream.of(Direction.VALUES) // Leaf - paw optimization - cache direction values
             .map(
                 direction -> {
                     StatePropertiesPredicate.Builder builder = StatePropertiesPredicate.Builder.properties()
diff --git a/net/minecraft/data/loot/BlockLootSubProvider.java b/net/minecraft/data/loot/BlockLootSubProvider.java
index 664425068cc6032df2ddfa7e7e9806bdd2f0d96e..92e88bccfc69dedcc49555243f33dd1194f62434 100644
--- a/net/minecraft/data/loot/BlockLootSubProvider.java
+++ b/net/minecraft/data/loot/BlockLootSubProvider.java
@@ -489,7 +489,7 @@ public abstract class BlockLootSubProvider implements LootTableSubProvider {
                             LootItem.lootTableItem(block)
                                 .when(builder)
                                 .apply(
-                                    Direction.values(),
+                                    Direction.VALUES, // Leaf - paw optimization - cache direction values
                                     direction -> SetItemCountFunction.setCount(ConstantValue.exactly(1.0F), true)
                                         .when(
                                             LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
@@ -513,7 +513,7 @@ public abstract class BlockLootSubProvider implements LootTableSubProvider {
                             block,
                             LootItem.lootTableItem(block)
                                 .apply(
-                                    Direction.values(),
+                                    Direction.VALUES, // Leaf - paw optimization - cache direction values
                                     direction -> SetItemCountFunction.setCount(ConstantValue.exactly(1.0F), true)
                                         .when(
                                             LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 5bfdb48ff7ef740a5a01d02e87f0264e17338de4..4e16552680f86cf1589f862cb38c0f43df344e6a 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -620,13 +620,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             if (!(this.packetListener instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener)
                 || loginPacketListener.state != net.minecraft.server.network.ServerLoginPacketListenerImpl.State.VERIFYING
                 || Connection.joinAttemptsThisTick++ < MAX_PER_TICK) {
-            // Paper start - detailed watchdog information
-            net.minecraft.network.PacketProcessor.packetProcessing.push(this.packetListener);
-            try {
             tickablePacketListener.tick();
-            } finally {
-                net.minecraft.network.PacketProcessor.packetProcessing.pop();
-            } // Paper end - detailed watchdog information
             } // Paper end - Buffer joins to world
         }
 
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 3b2a5d2c5d6afa8c8cbfe94464bd198e2d0c50f2..73c69215b57858c61474295749bd99a6fb8378d6 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -639,8 +639,12 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
             if (org.dreeam.leaf.config.modules.opt.OptimizeMobSpawning.enabled) { this.level.natureSpawnChunkMap.tick(this.level, list); } else { this.chunkMap.collectSpawningChunks(list); } // Leaf - optimize mob spawning
             profiler.popPush("shuffleSpawningChunks");
             // Paper start - chunk tick iteration optimisation
-            this.shuffleRandom.setSeed(this.level.random.nextLong());
-            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(list, this.shuffleRandom); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            // Leaf start - paw optimization - Only set seed if is really used
+            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
+                this.shuffleRandom.setSeed(this.level.random.nextLong());
+                Util.shuffle(list, this.shuffleRandom); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            }
+            // Leaf end - paw optimization - Only set seed if is really used
             // Paper end - chunk tick iteration optimisation
             profiler.popPush("tickSpawningChunks");
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 7028fce7b114631d9361c8598d9fc0a47653d6f7..2643dc54f899420b6847d1c537d0f1dd4a3c2e05 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1593,26 +1593,8 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
 
     }
 
-    // Paper start - log detailed entity tick information
-    // TODO replace with varhandle
-    static final java.util.concurrent.atomic.AtomicReference<Entity> currentlyTickingEntity = new java.util.concurrent.atomic.AtomicReference<>();
-
-    public static List<Entity> getCurrentlyTickingEntities() {
-        Entity ticking = currentlyTickingEntity.get();
-        List<Entity> ret = java.util.Arrays.asList(ticking == null ? me.titaniumtown.ArrayConstants.emptyEntityArray : new Entity[] { ticking }); // Gale - JettPack - reduce array allocations
-
-        return ret;
-    }
-    // Paper end - log detailed entity tick information
-
     public void tickNonPassenger(Entity entity) {
-        // Paper start - log detailed entity tick information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot tick an entity off-main");
-        try {
-            if (currentlyTickingEntity.get() == null) {
-                currentlyTickingEntity.lazySet(entity);
-            }
-            // Paper end - log detailed entity tick information
         entity.setOldPosAndRot();
         ProfilerFiller profilerFiller = Profiler.get();
         entity.tickCount++;
@@ -1629,13 +1611,6 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         for (Entity entity1 : entity.getPassengers()) {
             this.tickPassenger(entity, entity1, isActive); // Paper - EAR 2
         }
-        // Paper start - log detailed entity tick information
-        } finally {
-            if (currentlyTickingEntity.get() == entity) {
-                currentlyTickingEntity.lazySet(null);
-            }
-        }
-        // Paper end - log detailed entity tick information
     }
 
     private void tickPassenger(Entity ridingEntity, Entity passengerEntity, final boolean isActive) { // Paper - EAR 2
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index a0a3e5c32e216b7ffc3f90ccd5e513e3b980abc7..3212d952bbceb1f4c53e9994a97cdc73eaaa12e7 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1162,16 +1162,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return this.onGround;
     }
 
-    // Paper start - detailed watchdog information
-    public final Object posLock = new Object(); // Paper - log detailed entity tick information
-
-    @Nullable
-    private Vec3 moveVector;
-    private double moveStartX;
-    private double moveStartY;
-    private double moveStartZ;
-    // Paper end - detailed watchdog information
-
     public void move(MoverType type, Vec3 movement) {
         // Gale start - VMP - skip entity move if movement is zero
         if (!this.boundingBoxChanged && movement.equals(Vec3.ZERO)) {
@@ -1180,16 +1170,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         boundingBoxChanged = false;
         // Gale end - VMP - skip entity move if movement is zero
         final Vec3 originalMovement = movement; // Paper - Expose pre-collision velocity
-        // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
-        synchronized (this.posLock) {
-            this.moveStartX = this.getX();
-            this.moveStartY = this.getY();
-            this.moveStartZ = this.getZ();
-            this.moveVector = movement;
-        }
-        try {
-        // Paper end - detailed watchdog information
         if (this.noPhysics) {
             this.setPos(this.getX() + movement.x, this.getY() + movement.y, this.getZ() + movement.z);
             this.horizontalCollision = false;
@@ -1326,13 +1307,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                 profilerFiller.pop();
             }
         }
-        // Paper start - detailed watchdog information
-        } finally {
-            synchronized (this.posLock) { // Paper
-                this.moveVector = null;
-            } // Paper
-        }
-        // Paper end - detailed watchdog information
     }
 
     private void applyMovementEmissionAndPlaySound(Entity.MovementEmission movementEmission, Vec3 movement, BlockPos pos, BlockState state) {
@@ -5099,9 +5073,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     public void setDeltaMovement(Vec3 deltaMovement) {
         if (deltaMovement.isFinite()) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.deltaMovement = deltaMovement;
-            } // Paper - detailed watchdog information
         }
     }
 
@@ -5202,9 +5174,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         }
         // Paper end - Block invalid positions and bounding box
         if (this.position.x != x || this.position.y != y || this.position.z != z) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.position = new Vec3(x, y, z);
-            } // Paper - detailed watchdog information
             int floor = Mth.floor(x);
             int floor1 = Mth.floor(y);
             int floor2 = Mth.floor(z);
diff --git a/net/minecraft/world/entity/monster/Ghast.java b/net/minecraft/world/entity/monster/Ghast.java
index 667e51bdd9bf54bec9bcf5fa9a80b16ddbd99131..d6ff0c24d020942a5ed1e2b0b046d4d043ff33af 100644
--- a/net/minecraft/world/entity/monster/Ghast.java
+++ b/net/minecraft/world/entity/monster/Ghast.java
@@ -522,7 +522,7 @@ public class Ghast extends Mob implements Enemy {
                 if (!level.getBlockState(blockPos).isAir()) {
                     return false;
                 } else {
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         for (int i = 1; i < distanceToBlocks; i++) {
                             BlockPos blockPos1 = blockPos.relative(direction, i);
                             if (!level.getBlockState(blockPos1).isAir()) {
diff --git a/net/minecraft/world/entity/monster/Shulker.java b/net/minecraft/world/entity/monster/Shulker.java
index 2b73382e5bcf47da483ea88e2eae480ebb48e8f9..12e3ba1a545d5ec208ed33af66429f56fcdbed7f 100644
--- a/net/minecraft/world/entity/monster/Shulker.java
+++ b/net/minecraft/world/entity/monster/Shulker.java
@@ -398,7 +398,7 @@ public class Shulker extends AbstractGolem implements Enemy {
     }
 
     protected @Nullable Direction findAttachableSurface(BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.canStayAt(pos, direction)) {
                 return direction;
             }
diff --git a/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java b/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
index f178c45c834aa276522c4ecd6995d9eaf63d3eb6..95609d97e91c55fd289d1a2b10258af4c8f52779 100644
--- a/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
+++ b/net/minecraft/world/level/block/BaseCoralPlantTypeBlock.java
@@ -43,7 +43,7 @@ public abstract class BaseCoralPlantTypeBlock extends Block implements SimpleWat
         if (state.getValue(WATERLOGGED)) {
             return true;
         } else {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (level.getFluidState(pos.relative(direction)).is(FluidTags.WATER)) {
                     return true;
                 }
diff --git a/net/minecraft/world/level/block/BaseFireBlock.java b/net/minecraft/world/level/block/BaseFireBlock.java
index 42e897dc4bee8e9c075415d494ce13c404f2042a..8e63eb08cd1bd1c0d0aef8952d5f5cf1dd0f9d8b 100644
--- a/net/minecraft/world/level/block/BaseFireBlock.java
+++ b/net/minecraft/world/level/block/BaseFireBlock.java
@@ -207,7 +207,7 @@ public abstract class BaseFireBlock extends Block {
             BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
             boolean flag = false;
 
-            for (Direction direction1 : Direction.values()) {
+            for (Direction direction1 : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (PortalShape.FRAME.test(level.getBlockState(mutableBlockPos.set(pos).move(direction1)), level, mutableBlockPos)) { // Purpur - Crying obsidian valid for portal frames
                     flag = true;
                     break;
diff --git a/net/minecraft/world/level/block/ConcretePowderBlock.java b/net/minecraft/world/level/block/ConcretePowderBlock.java
index 0284c631e8e2b303f4f95927d99513423dad4b0c..3a7e14dd369aeb214a800c980c4c1c11fffae906 100644
--- a/net/minecraft/world/level/block/ConcretePowderBlock.java
+++ b/net/minecraft/world/level/block/ConcretePowderBlock.java
@@ -75,7 +75,7 @@ public class ConcretePowderBlock extends FallingBlock {
         boolean flag = false;
         BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockState blockState = level.getBlockState(mutableBlockPos);
             if (direction != Direction.DOWN || canSolidify(blockState)) {
                 mutableBlockPos.setWithOffset(pos, direction);
diff --git a/net/minecraft/world/level/block/CoralBlock.java b/net/minecraft/world/level/block/CoralBlock.java
index 337985a30767aaac573fce6301b2690473f8305a..bea40c819f479c13c351144533096a14237d1c1f 100644
--- a/net/minecraft/world/level/block/CoralBlock.java
+++ b/net/minecraft/world/level/block/CoralBlock.java
@@ -66,7 +66,7 @@ public class CoralBlock extends Block {
 
     protected boolean scanForWater(BlockGetter level, BlockPos pos) {
         if (!((net.minecraft.world.level.LevelAccessor) level).getMinecraftWorld().purpurConfig.coralDieOutsideWater) return true; // Purpur - Config to not let coral die
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             FluidState fluidState = level.getFluidState(pos.relative(direction));
             if (fluidState.is(FluidTags.WATER)) {
                 return true;
diff --git a/net/minecraft/world/level/block/CreakingHeartBlock.java b/net/minecraft/world/level/block/CreakingHeartBlock.java
index dfb9a10622ba2a0c2759bb563fefd729f1109865..1b107d778d4a5a7783f3a123711045170fac7e7a 100644
--- a/net/minecraft/world/level/block/CreakingHeartBlock.java
+++ b/net/minecraft/world/level/block/CreakingHeartBlock.java
@@ -126,7 +126,7 @@ public class CreakingHeartBlock extends BaseEntityBlock {
     }
 
     private static boolean isSurroundedByLogs(LevelAccessor level, BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos blockPos = pos.relative(direction);
             BlockState blockState = level.getBlockState(blockPos);
             if (!blockState.is(BlockTags.PALE_OAK_LOGS)) {
diff --git a/net/minecraft/world/level/block/DiodeBlock.java b/net/minecraft/world/level/block/DiodeBlock.java
index 02ffb5569e2405d86b3a4a695dd17c9372169ff7..4be74d72f60ca6f1792ec5d1bc24ec6bd567e28d 100644
--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -99,7 +99,7 @@ public abstract class DiodeBlock extends HorizontalDirectionalBlock {
             dropResources(state, level, pos, blockEntity);
             level.removeBlock(pos, false);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
         }
diff --git a/net/minecraft/world/level/block/FireBlock.java b/net/minecraft/world/level/block/FireBlock.java
index 14a4189166b862b66a8383046f2abfe02ccbccd8..9219856e9594466f0b31d9463ce416334299bdfe 100644
--- a/net/minecraft/world/level/block/FireBlock.java
+++ b/net/minecraft/world/level/block/FireBlock.java
@@ -136,7 +136,7 @@ public class FireBlock extends BaseFireBlock {
         if (!this.canBurn(blockState) && !blockState.isFaceSturdy(level, blockPos, Direction.UP)) {
             BlockState blockState1 = this.defaultBlockState();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 BooleanProperty booleanProperty = PROPERTY_BY_DIRECTION.get(direction);
                 if (booleanProperty != null) {
                     blockState1 = blockState1.setValue(booleanProperty, this.canBurn(level.getBlockState(pos.relative(direction))));
@@ -306,7 +306,7 @@ public class FireBlock extends BaseFireBlock {
     }
 
     private boolean isValidFireLocation(BlockGetter level, BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.canBurn(level.getBlockState(pos.relative(direction)))) {
                 return true;
             }
@@ -321,7 +321,7 @@ public class FireBlock extends BaseFireBlock {
         } else {
             int i = 0;
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 BlockState blockState = level.getBlockState(pos.relative(direction));
                 i = Math.max(this.getIgniteOdds(blockState), i);
             }
diff --git a/net/minecraft/world/level/block/FrostedIceBlock.java b/net/minecraft/world/level/block/FrostedIceBlock.java
index 4d62fa1c74ee3b810f02f23cc18ee4e887e93910..c6b3b37aee416ad6a3c22b027c8a9aaab336eedd 100644
--- a/net/minecraft/world/level/block/FrostedIceBlock.java
+++ b/net/minecraft/world/level/block/FrostedIceBlock.java
@@ -49,7 +49,7 @@ public class FrostedIceBlock extends IceBlock {
             if (i > 11 - state.getValue(AGE) - state.getLightBlock() && this.slightlyMelt(state, level, pos)) {
                 BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-                for (Direction direction : Direction.values()) {
+                for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     mutableBlockPos.setWithOffset(pos, direction);
                     BlockState blockState = level.getBlockState(mutableBlockPos);
                     if (blockState.is(this) && !this.slightlyMelt(blockState, level, mutableBlockPos)) {
@@ -88,7 +88,7 @@ public class FrostedIceBlock extends IceBlock {
         int i = 0;
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             mutableBlockPos.setWithOffset(pos, direction);
             if (level.getBlockState(mutableBlockPos).is(this)) {
                 if (++i >= neighborsRequired) {
diff --git a/net/minecraft/world/level/block/LeavesBlock.java b/net/minecraft/world/level/block/LeavesBlock.java
index d5b114dca7b28d910f03ae3468e891d5d9863897..c3562b6747e2a495bddf2987202a1a9b1f639025 100644
--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -120,7 +120,7 @@ public abstract class LeavesBlock extends Block implements SimpleWaterloggedBloc
         int i = 7;
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             mutableBlockPos.setWithOffset(pos, direction);
             i = Math.min(i, getDistanceAt(level.getBlockState(mutableBlockPos)) + 1);
             if (i == 1) {
diff --git a/net/minecraft/world/level/block/MultifaceBlock.java b/net/minecraft/world/level/block/MultifaceBlock.java
index 2e185de23478010caa8d0063ae8521739a10932f..3e5d0bf2ecc70e5f3a765cea0fb99199c2257936 100644
--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -73,7 +73,7 @@ public class MultifaceBlock extends Block implements SimpleWaterloggedBlock {
         } else {
             Set<Direction> set = EnumSet.noneOf(Direction.class);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (hasFace(state, direction)) {
                     set.add(direction);
                 }
@@ -86,7 +86,7 @@ public class MultifaceBlock extends Block implements SimpleWaterloggedBlock {
     public static Set<Direction> unpack(byte packedDirections) {
         Set<Direction> set = EnumSet.noneOf(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if ((packedDirections & (byte)(1 << direction.ordinal())) > 0) {
                 set.add(direction);
             }
diff --git a/net/minecraft/world/level/block/RedStoneOreBlock.java b/net/minecraft/world/level/block/RedStoneOreBlock.java
index 8ef666854fa5342261626ad0bb12d72d1cb0cee2..bd39df675e39e5e5005db74e05d9f5e111868c31 100644
--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -134,7 +134,7 @@ public class RedStoneOreBlock extends Block {
         double d = 0.5625;
         RandomSource randomSource = level.random;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos blockPos = pos.relative(direction);
             if (!level.getBlockState(blockPos).isSolidRender()) {
                 Direction.Axis axis = direction.getAxis();
diff --git a/net/minecraft/world/level/block/RedStoneWireBlock.java b/net/minecraft/world/level/block/RedStoneWireBlock.java
index ebc43f02c356e94d889e20ed26f66dc216b305a6..95dcef271566314e90edf64c271715359eabac1d 100644
--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -359,7 +359,7 @@ public class RedStoneWireBlock extends Block {
         if (level.getBlockState(pos).is(this)) {
             level.updateNeighborsAt(pos, this);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
         }
@@ -387,7 +387,7 @@ public class RedStoneWireBlock extends Block {
     @Override
     protected void affectNeighborsAfterRemoval(BlockState state, ServerLevel level, BlockPos pos, boolean movedByPiston) {
         if (!movedByPiston) {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
diff --git a/net/minecraft/world/level/block/RedstoneTorchBlock.java b/net/minecraft/world/level/block/RedstoneTorchBlock.java
index 5365cc6c2ec7f86376bf27551493b6621d4c412e..17715d23610ace2c4051277d02e405e079387f51 100644
--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -48,7 +48,7 @@ public class RedstoneTorchBlock extends BaseTorchBlock {
     private void notifyNeighbors(Level level, BlockPos pos, BlockState state) {
         Orientation orientation = this.randomOrientation(level, state);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             level.updateNeighborsAt(pos.relative(direction), this, ExperimentalRedstoneUtils.withFront(orientation, direction));
         }
     }
diff --git a/net/minecraft/world/level/block/SculkSensorBlock.java b/net/minecraft/world/level/block/SculkSensorBlock.java
index 80e6ba8f026a6565696f18304507b2d0a347650c..40e38007a07737efd167fd0f4a78f2ca33bed8bd 100644
--- a/net/minecraft/world/level/block/SculkSensorBlock.java
+++ b/net/minecraft/world/level/block/SculkSensorBlock.java
@@ -253,7 +253,7 @@ public class SculkSensorBlock extends BaseEntityBlock implements SimpleWaterlogg
     }
 
     public static void tryResonateVibration(@Nullable Entity entity, Level level, BlockPos pos, int frequency) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             BlockPos blockPos = pos.relative(direction);
             BlockState blockState = level.getBlockState(blockPos);
             if (blockState.is(BlockTags.VIBRATION_RESONATORS)) {
diff --git a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
index 63ec7706e52f1951a50aa0850c07fb63800d9a74..f5301c69352648581050538344cb477b062d9226 100644
--- a/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CreakingHeartBlockEntity.java
@@ -250,7 +250,7 @@ public class CreakingHeartBlockEntity extends BlockEntity {
     private Optional<BlockPos> spreadResin(ServerLevel level) {
         Mutable<BlockPos> mutable = new MutableObject<>(null);
         BlockPos.breadthFirstTraversal(this.worldPosition, 2, 64, (pos, output) -> {
-            for (Direction direction : Util.shuffledCopy(Direction.values(), level.random)) {
+            for (Direction direction : Util.shuffledCopy(Direction.VALUES, level.random)) { // Leaf - paw optimization - cache direction values
                 BlockPos blockPos = pos.relative(direction);
                 if (level.getBlockState(blockPos).is(BlockTags.PALE_OAK_LOGS)) {
                     output.accept(blockPos);
@@ -260,7 +260,7 @@ public class CreakingHeartBlockEntity extends BlockEntity {
             if (!level.getBlockState(pos).is(BlockTags.PALE_OAK_LOGS)) {
                 return BlockPos.TraversalNodeStatus.ACCEPT;
             } else {
-                for (Direction direction : Util.shuffledCopy(Direction.values(), level.random)) {
+                for (Direction direction : Util.shuffledCopy(Direction.VALUES, level.random)) { // Leaf - paw optimization - cache direction values
                     BlockPos blockPos = pos.relative(direction);
                     BlockState blockState = level.getBlockState(blockPos);
                     Direction opposite = direction.getOpposite();
diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 0fdf7647070c2e8c9ce8a41c7d3e1e38f6b2e4c9..02eca48e5784557da317e20f23c68b2989799849 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -256,7 +256,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
     public int getParticleAmount() {
         int i = 0;
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             i += this.shouldRenderFace(direction) ? 1 : 0;
         }
 
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 3bbfa079a2a2da3de361352738b6101894acf82c..1576f07cbb3708c9878dcf81d331a04ec0e0d9ce 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -128,7 +128,7 @@ public class PistonBaseBlock extends DirectionalBlock {
     }
 
     private boolean getNeighborSignal(SignalGetter level, BlockPos pos, Direction direction) {
-        for (Direction direction1 : Direction.values()) {
+        for (Direction direction1 : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction1 != direction && level.hasSignal(pos.relative(direction1), direction1)) {
                 return true;
             }
@@ -139,7 +139,7 @@ public class PistonBaseBlock extends DirectionalBlock {
         } else {
             BlockPos blockPos = pos.above();
 
-            for (Direction direction2 : Direction.values()) {
+            for (Direction direction2 : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (direction2 != Direction.DOWN && level.hasSignal(blockPos.relative(direction2), direction2)) {
                     return true;
                 }
diff --git a/net/minecraft/world/level/block/piston/PistonStructureResolver.java b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
index 5a3660e02bc805e9a35a81b8a61f07b3f20b5ba9..05b0440e9b11b44fb039aabf57be556dd64e0b8b 100644
--- a/net/minecraft/world/level/block/piston/PistonStructureResolver.java
+++ b/net/minecraft/world/level/block/piston/PistonStructureResolver.java
@@ -168,7 +168,7 @@ public class PistonStructureResolver {
     private boolean addBranchingBlocks(BlockPos fromPos) {
         BlockState blockState = this.level.getBlockState(fromPos);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (direction.getAxis() != this.pushDirection.getAxis()) {
                 BlockPos blockPos = fromPos.relative(direction);
                 BlockState blockState1 = this.level.getBlockState(blockPos);
diff --git a/net/minecraft/world/level/block/state/pattern/BlockPattern.java b/net/minecraft/world/level/block/state/pattern/BlockPattern.java
index f834d487136627c07c893b8d68a6929ed46f20b9..e6480199995e44987238ca345b7de9f36c90f147 100644
--- a/net/minecraft/world/level/block/state/pattern/BlockPattern.java
+++ b/net/minecraft/world/level/block/state/pattern/BlockPattern.java
@@ -76,8 +76,8 @@ public class BlockPattern {
         int max = Math.max(Math.max(this.width, this.height), this.depth);
 
         for (BlockPos blockPos : BlockPos.betweenClosed(pos, pos.offset(max - 1, max - 1, max - 1))) {
-            for (Direction direction : Direction.values()) {
-                for (Direction direction1 : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
+                for (Direction direction1 : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     if (direction1 != direction && direction1 != direction.getOpposite()) {
                         BlockPattern.BlockPatternMatch blockPatternMatch = this.matches(blockPos, direction, direction1, loadingCache);
                         if (blockPatternMatch != null) {
diff --git a/net/minecraft/world/level/chunk/UpgradeData.java b/net/minecraft/world/level/chunk/UpgradeData.java
index acd1d7b7987a1674cc07e2b1e546721254c4c676..7f95747569430161e6b7f23e0ce2a89367f1c097 100644
--- a/net/minecraft/world/level/chunk/UpgradeData.java
+++ b/net/minecraft/world/level/chunk/UpgradeData.java
@@ -146,14 +146,13 @@ public class UpgradeData {
             int i3 = pos.getMinBlockX() + (!flag4 || !flag3 && !flag2 ? (flag1 ? 0 : 15) : 14);
             int i4 = pos.getMinBlockZ() + (!flag4 || !flag && !flag1 ? (flag3 ? 0 : 15) : 1);
             int i5 = pos.getMinBlockZ() + (!flag4 || !flag && !flag1 ? (flag3 ? 0 : 15) : 14);
-            Direction[] directions1 = Direction.values();
             BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
             for (BlockPos blockPos : BlockPos.betweenClosed(i2, level.getMinY(), i4, i3, level.getMaxY(), i5)) {
                 BlockState blockState = level.getBlockState(blockPos);
                 BlockState blockState1 = blockState;
 
-                for (Direction direction : directions1) {
+                for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                     mutableBlockPos.setWithOffset(blockPos, direction);
                     blockState1 = updateState(blockState1, direction, level, blockPos, mutableBlockPos);
                 }
@@ -179,7 +178,6 @@ public class UpgradeData {
             int[] ints = this.index[i];
             this.index[i] = null;
             if (ints != null && ints.length > 0) {
-                Direction[] directions = Direction.values();
                 PalettedContainer<BlockState> states = section.getStates();
                 int sectionYFromSectionIndex = chunk.getSectionYFromSectionIndex(i);
                 int blockPosCoord = SectionPos.sectionToBlockCoord(sectionYFromSectionIndex);
@@ -192,7 +190,7 @@ public class UpgradeData {
                     BlockState blockState = states.get(i1);
                     BlockState blockState1 = blockState;
 
-                    for (Direction direction : directions) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         mutableBlockPos1.setWithOffset(mutableBlockPos, direction);
                         if (SectionPos.blockToSectionCoord(mutableBlockPos.getX()) == pos.x && SectionPos.blockToSectionCoord(mutableBlockPos.getZ()) == pos.z) {
                             blockState1 = updateState(blockState1, direction, level, mutableBlockPos, mutableBlockPos1);
diff --git a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
index 00a8e824c893593f6cc4221a14dba7952e21443f..3f5ffa8679c6ba6684175938d41d5a947063ebf6 100644
--- a/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
+++ b/net/minecraft/world/level/gameevent/vibrations/VibrationSystem.java
@@ -254,7 +254,7 @@ public interface VibrationSystem {
             Vec3 vec3 = new Vec3(Mth.floor(eventPos.x) + 0.5, Mth.floor(eventPos.y) + 0.5, Mth.floor(eventPos.z) + 0.5);
             Vec3 vec31 = new Vec3(Mth.floor(vibrationUserPos.x) + 0.5, Mth.floor(vibrationUserPos.y) + 0.5, Mth.floor(vibrationUserPos.z) + 0.5);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 Vec3 vec32 = vec3.relative(direction, 1.0E-5F);
                 if (level.isBlockInLine(new ClipBlockStateContext(vec32, vec31, state -> state.is(BlockTags.OCCLUDES_VIBRATION_SIGNALS))).getType()
                     != HitResult.Type.BLOCK) {
diff --git a/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java b/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
index 04aaaa6e3a6b6250fbe609be36a3c0927e217566..1cc0c8ba49175c320f8873c77d1b21fc0c48800e 100644
--- a/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BlueIceFeature.java
@@ -27,7 +27,7 @@ public class BlueIceFeature extends Feature<NoneFeatureConfiguration> {
         } else {
             boolean flag = false;
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (direction != Direction.DOWN && worldGenLevel.getBlockState(blockPos.relative(direction)).is(Blocks.PACKED_ICE)) {
                     flag = true;
                     break;
@@ -52,7 +52,7 @@ public class BlueIceFeature extends Feature<NoneFeatureConfiguration> {
                         );
                         BlockState blockState = worldGenLevel.getBlockState(blockPos1);
                         if (blockState.isAir() || blockState.is(Blocks.WATER) || blockState.is(Blocks.PACKED_ICE) || blockState.is(Blocks.ICE)) {
-                            for (Direction direction1 : Direction.values()) {
+                            for (Direction direction1 : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                                 BlockState blockState1 = worldGenLevel.getBlockState(blockPos1.relative(direction1));
                                 if (blockState1.is(Blocks.BLUE_ICE)) {
                                     worldGenLevel.setBlock(blockPos1, Blocks.BLUE_ICE.defaultBlockState(), Block.UPDATE_CLIENTS);
diff --git a/net/minecraft/world/level/levelgen/feature/Feature.java b/net/minecraft/world/level/levelgen/feature/Feature.java
index c6d7a686bcc21d041444b647deeab2acf4859194..55c4b6d49fb01f5d4f631b825ffc248ac42061b7 100644
--- a/net/minecraft/world/level/levelgen/feature/Feature.java
+++ b/net/minecraft/world/level/levelgen/feature/Feature.java
@@ -207,7 +207,7 @@ public abstract class Feature<FC extends FeatureConfiguration> {
     public static boolean checkNeighbors(Function<BlockPos, BlockState> adjacentStateAccessor, BlockPos pos, Predicate<BlockState> filter) {
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             mutableBlockPos.setWithOffset(pos, direction);
             if (filter.test(adjacentStateAccessor.apply(mutableBlockPos))) {
                 return true;
diff --git a/net/minecraft/world/level/levelgen/feature/GlowstoneFeature.java b/net/minecraft/world/level/levelgen/feature/GlowstoneFeature.java
index 7b5a375b48ab8ec778dfec1bf53b115da4387dc1..6504564e4102829744e5f853c48e58a212442bc4 100644
--- a/net/minecraft/world/level/levelgen/feature/GlowstoneFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/GlowstoneFeature.java
@@ -36,7 +36,7 @@ public class GlowstoneFeature extends Feature<NoneFeatureConfiguration> {
                     if (worldGenLevel.getBlockState(blockPos1).isAir()) {
                         int i1 = 0;
 
-                        for (Direction direction : Direction.values()) {
+                        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                             if (worldGenLevel.getBlockState(blockPos1.relative(direction)).is(Blocks.GLOWSTONE)) {
                                 i1++;
                             }
diff --git a/net/minecraft/world/level/levelgen/feature/TreeFeature.java b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
index 0a8b1925bc8cbd0c41444835dac6a76a63ffe14f..b86d41bfea27833e37a87847bf7ec5cbd40ae2bc 100644
--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -205,7 +205,7 @@ public class TreeFeature extends Feature<TreeConfiguration> {
 
                     discreteVoxelShape.fill(blockPos1.getX() - box.minX(), blockPos1.getY() - box.minY(), blockPos1.getZ() - box.minZ());
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         mutableBlockPos.setWithOffset(blockPos1, direction);
                         if (box.isInside(mutableBlockPos)) {
                             int i3 = mutableBlockPos.getX() - box.minX();
diff --git a/net/minecraft/world/level/levelgen/feature/VinesFeature.java b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
index ba477a03fe35bca6710a35ea20f332ff937ad132..93ce4f8e404c0abf1cc1503b1e0c4e3b32a02d5f 100644
--- a/net/minecraft/world/level/levelgen/feature/VinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/VinesFeature.java
@@ -22,7 +22,7 @@ public class VinesFeature extends Feature<NoneFeatureConfiguration> {
         if (!worldGenLevel.isEmptyBlock(blockPos)) {
             return false;
         } else {
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 if (direction != Direction.DOWN && VineBlock.isAcceptableNeighbour(worldGenLevel, blockPos.relative(direction), direction)) {
                     worldGenLevel.setBlock(
                         blockPos, Blocks.VINE.defaultBlockState().setValue(VineBlock.getPropertyForFace(direction), true), Block.UPDATE_CLIENTS
diff --git a/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java b/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
index 8763d03382ee33b8024bc390b30f5c3e55f26d97..0f9cab559d4652bfce635481d22dc2035552b91e 100644
--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
@@ -38,7 +38,7 @@ public class CreakingHeartDecorator extends TreeDecorator {
                 List<BlockPos> list1 = new ArrayList<>(list);
                 Util.shuffle(list1, randomSource);
                 Optional<BlockPos> optional = list1.stream().filter(pos -> {
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         if (!context.checkBlock(pos.relative(direction), blockState -> blockState.is(BlockTags.LOGS))) {
                             return false;
                         }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java b/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
index 57d3c7a82917a8ab66135993e9864c7fb6e4d493..5a53cf2380b00dd29968e533ccefdda4de2968c2 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/BuriedTreasurePieces.java
@@ -55,7 +55,7 @@ public class BuriedTreasurePieces {
                     || blockState1 == Blocks.DIORITE.defaultBlockState()) {
                     BlockState blockState2 = !blockState.isAir() && !this.isLiquid(blockState) ? blockState : Blocks.SAND.defaultBlockState();
 
-                    for (Direction direction : Direction.values()) {
+                    for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                         BlockPos blockPos = mutableBlockPos.relative(direction);
                         BlockState blockState3 = level.getBlockState(blockPos);
                         if (blockState3.isAir() || this.isLiquid(blockState3)) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
index 2eb836f3bf435b4d66dba9bab9a2b718b06bfa62..8e523cb32e471083390aa9764d7f1e515f976e52 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidStructure.java
@@ -49,7 +49,7 @@ public class DesertPyramidStructure extends SinglePieceStructure {
             }
         }
 
-        ObjectArrayList<BlockPos> list = new ObjectArrayList<>(set.stream().toList());
+        ObjectArrayList<BlockPos> list = new ObjectArrayList<>(set); // Leaf - paw optimization - TODO: use array
         RandomSource randomSource = RandomSource.create(level.getSeed()).forkPositional().at(pieces.calculateBoundingBox().getCenter());
         Util.shuffle(list, randomSource);
         int min = Math.min(set.size(), randomSource.nextInt(5, 8));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
index d6d3e20e5425c5701e829b1dc71d844b71cd0e34..d19e0b27fbc850eeb05571192724842a5c1d40ce 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -560,7 +560,7 @@ public class MineshaftPieces {
             BlockPos.MutableBlockPos worldPos = this.getWorldPos(x, y, z);
             int i = 0;
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 worldPos.move(direction);
                 if (box.isInside(worldPos) && level.getBlockState(worldPos).isFaceSturdy(level, worldPos, direction.getOpposite())) {
                     if (++i >= required) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
index e24b8556dce62bc38d4e22beb15276a5f8475ad7..7fff2f3bd6a0f082bca0b638786ab035afbb9c82 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -244,7 +244,7 @@ public class OceanMonumentPieces {
                     for (int i2 = 0; i2 < 3; i2++) {
                         int roomIndex = getRoomIndex(i, i2, i1);
                         if (roomDefinitions[roomIndex] != null) {
-                            for (Direction direction : Direction.values()) {
+                            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                                 int i3 = i + direction.getStepX();
                                 int i4 = i2 + direction.getStepY();
                                 int i5 = i1 + direction.getStepZ();
diff --git a/net/minecraft/world/level/material/LavaFluid.java b/net/minecraft/world/level/material/LavaFluid.java
index be35248556e21b227adde0f02f7f958d7c860fa2..e6e46a8bf3fb9fd75fb8074993403f89dea271fc 100644
--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -140,7 +140,7 @@ public abstract class LavaFluid extends FlowingFluid {
     }
 
     private boolean hasFlammableNeighbours(LevelReader level, BlockPos pos) {
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             if (this.isFlammable(level, pos.relative(direction))) {
                 return true;
             }
diff --git a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
index ea90ada690ba23aeb6990d828b4b3513798ce21e..915536f109712ccd4f6c20a5667817e183f226f3 100644
--- a/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/AmphibiousNodeEvaluator.java
@@ -97,7 +97,7 @@ public class AmphibiousNodeEvaluator extends WalkNodeEvaluator {
         if (pathTypeFromState == PathType.WATER) {
             BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 mutableBlockPos.set(x, y, z).move(direction);
                 PathType pathTypeFromState1 = context.getPathTypeFromState(mutableBlockPos.getX(), mutableBlockPos.getY(), mutableBlockPos.getZ());
                 if (pathTypeFromState1 == PathType.BLOCKED) {
diff --git a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
index f61f83e7f6541a4354febcef368c096c6d5a36eb..16550a6ed9c22104328f0ed017f437a8c0bcc6f7 100644
--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -51,7 +51,7 @@ public class SwimNodeEvaluator extends NodeEvaluator {
         int i = 0;
         Map<Direction, Node> map = Maps.newEnumMap(Direction.class);
 
-        for (Direction direction : Direction.values()) {
+        for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
             Node node1 = this.findAcceptedNode(node.x + direction.getStepX(), node.y + direction.getStepY(), node.z + direction.getStepZ());
             map.put(direction, node1);
             if (this.isNodeValid(node1)) {
diff --git a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
index 53b8df8e6dd379126af3e1feb2220e222430ce94..cd837af838481d104457d0eabfe9a290e89d67d0 100644
--- a/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/DefaultRedstoneWireEvaluator.java
@@ -35,7 +35,7 @@ public class DefaultRedstoneWireEvaluator extends RedstoneWireEvaluator {
             Set<BlockPos> set = Sets.newHashSet();
             set.add(pos);
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 set.add(pos.relative(direction));
             }
 
diff --git a/net/minecraft/world/level/redstone/Orientation.java b/net/minecraft/world/level/redstone/Orientation.java
index 1d2a1fd3f59a0b04e10a08fd2723737301d90755..42452291015121f4a457fc7838b4eee401449a5b 100644
--- a/net/minecraft/world/level/redstone/Orientation.java
+++ b/net/minecraft/world/level/redstone/Orientation.java
@@ -135,7 +135,7 @@ public class Orientation {
                 start.withSideBias.put(sideBias, generateContext(new Orientation(start.up, start.front, sideBias), output));
             }
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 Direction direction1 = start.up;
                 if (direction == start.up) {
                     direction1 = start.front.getOpposite();
@@ -148,7 +148,7 @@ public class Orientation {
                 start.withFront.put(direction, generateContext(new Orientation(direction1, direction, start.sideBias), output));
             }
 
-            for (Direction direction : Direction.values()) {
+            for (Direction direction : Direction.VALUES) { // Leaf - paw optimization - cache direction values
                 Direction direction1x = start.front;
                 if (direction == start.front) {
                     direction1x = start.up.getOpposite();
